pkg/sync/semaphore.go

```diff
@@ -0,0 +1,154 @@
+package sync
+
+import (
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+	sema "golang.org/x/sync/semaphore"
+)
+
+type PrioritySemaphore struct {
+	name       string
+	limit      int
+	pending    *priorityQueue
+	semaphore  *sema.Weighted
+	lockHolder map[string]bool
+	lock       *sync.Mutex
+	syncer     bool
+	logger     *zap.SugaredLogger
+}
+
+var _ Semaphore = &PrioritySemaphore{}
+
+func NewSemaphore(name string, limit int, logger *zap.SugaredLogger) *PrioritySemaphore {
+	return &PrioritySemaphore{
+		name:       name,
+		limit:      limit,
+		pending:    &priorityQueue{itemByKey: make(map[string]*item)},
```

Comment:
is this queue ephemeral? 
Say the app crashes (may be the node goes down) and the pod is restarted, how would this information be rebuilt?